package/src/lsif.ts

```diff
@@ -0,0 +1,229 @@
+import * as sourcegraph from 'sourcegraph'
+import * as LSP from 'vscode-languageserver-types'
+import { convertLocations, convertHover } from './lsp-conversion'
+import { queryGraphQL } from './api'
+
+function repositoryFromDoc(doc: sourcegraph.TextDocument): string {
+    const url = new URL(doc.uri)
+    return url.hostname + url.pathname
+}
+
+function commitFromDoc(doc: sourcegraph.TextDocument): string {
+    const url = new URL(doc.uri)
+    return url.search.slice(1)
+}
+
+function pathFromDoc(doc: sourcegraph.TextDocument): string {
+    const url = new URL(doc.uri)
+    return url.hash.slice(1)
+}
+
+function setPath(doc: sourcegraph.TextDocument, path: string): string {
+    const url = new URL(doc.uri)
+    url.hash = path
+    return url.href
+}
+
+async function queryLSIF({
+    doc,
+    method,
+    path,
+    position,
+}: {
+    doc: sourcegraph.TextDocument
+    method: string
+    path: string
+    position: LSP.Position
+}): Promise<any> {
+    const url = new URL(
+        '.api/lsif/request',
+        sourcegraph.internal.sourcegraphURL
+    )
+    url.searchParams.set('repository', repositoryFromDoc(doc))
+    url.searchParams.set('commit', commitFromDoc(doc))
+
+    const response = await fetch(url.href, {
+        method: 'POST',
+        headers: new Headers({
+            'content-type': 'application/json',
+            'x-requested-with': 'Sourcegraph LSIF extension',
+        }),
+        body: JSON.stringify({
+            method,
+            path,
+            position,
+        }),
+    })
+    if (!response.ok) {
+        throw new Error(`LSIF /request returned ${response.statusText}`)
+    }
+    return await response.json()
+}
+
+export const mkIsLSIFAvailable = (lsifDocs: Map<string, Promise<boolean>>) => (
+    doc: sourcegraph.TextDocument,
+    pos: sourcegraph.Position
+): Promise<boolean> => {
+    if (!sourcegraph.configuration.get().get('codeIntel.lsif')) {
+        return Promise.resolve(false)
+    }
+
+    if (lsifDocs.has(doc.uri)) {
+        return lsifDocs.get(doc.uri)!
+    }
+
+    const url = new URL('.api/lsif/exists', sourcegraph.internal.sourcegraphURL)
+    url.searchParams.set('repository', repositoryFromDoc(doc))
+    url.searchParams.set('commit', commitFromDoc(doc))
+    url.searchParams.set('file', pathFromDoc(doc))
+
+    const hasLSIFPromise = (async () => {
+        try {
+            // Prevent leaking the name of a private repository to
+            // Sourcegraph.com by relying on the Sourcegraph extension host's
+            // private repository detection, which will throw an error when
+            // making a GraphQL request.
```

Comment:
This part of the code is not clear to me.
- How would we leak private repo names? Wouldn't we always contact the private instance?
- Why would the extension host throw an error on any GraphQL request _to the private instance_? How would it even intercept GraphQL requests?

Reply:
Private repo names would have been leaked by a browser extension running on private GitHub.com repos.

There's already code in the extension host to detect if the page is a private repo. This logic triggers that code to get run at the expense of an extra GraphQL request. Do you have any ideas to improve this? Maybe the extension API could expose `isPrivateRepository`?